include/gunrock/applications/async/bfs.cuh

```diff
@@ -0,0 +1,185 @@
+      result_type& _result
+  ) : graph_slice(G), param(_param), result(_result) {}
+  
+  void init() {}
+  // </boiler-plate>
+  
+  void reset() {
+    // <user-defined>
+    auto g = this->get_graph();
+    auto n_vertices = g.get_number_of_vertices();
+    
+    auto single_source = param.single_source;
+    auto d_depth       = thrust::device_pointer_cast(this->result.depth);
+    thrust::fill(thrust::device, d_depth + 0, d_depth + n_vertices, n_vertices + 1);
+    thrust::fill(thrust::device, d_depth + single_source, d_depth + single_source + 1, 0);
+    // </user-defined>
+  }
+};
+
+// --
+// Enactor
+
+template<typename queue_t, typename val_t>
+__global__ void _push_one(queue_t q, val_t val) {
+    if(LANE_ == 0) q.push(val);
+}
+
+// This is very close to compatible w/ standard Gunrock enactor_t
+// However, it doesn't use the `context` argument, so not joining yet
+template<typename problem_t, typename single_queue_t=uint32_t>
+struct enactor_t {
+    using vertex_t = typename problem_t::vertex_t;
+    using edge_t   = typename problem_t::edge_t;
+    using queue_t  = MaxCountQueue::Queues<vertex_t, single_queue_t>;
+    
+    problem_t* problem;
+    queue_t q;
+    
+    int numBlock  = 56 * 5;
+    int numThread = 256;
+
+    // <boiler-plate>
+    enactor_t(
+      problem_t* _problem,
+      uint32_t  min_iter=800, 
+      int       num_queue=4
+    ) : problem(_problem) { 
+        
+        
+        auto n_vertices = problem->get_graph().get_number_of_vertices();
+        
+        auto capacity = min(
+          single_queue_t(1 << 30), 
+          max(single_queue_t(1024),  single_queue_t(n_vertices * 1.5))
+        );
+        
+        q.init(capacity, num_queue, min_iter);
+        q.reset();
+    }
+    // <boiler-plate>
+
+    // <user-defined>
+    void prepare_frontier() {
+      _push_one<<<1, 32>>>(q, problem->param.single_source);
+    }
+    // </user-defined>
+    
+    void enact() { // Is there some way to restructure this to follow the `loop` semantics?
+      
+      // <boiler-plate>
+      prepare_frontier();
+      // </boiler-plate>
+      
+      // <user-defined>
+      auto G        = problem->get_graph();
+      edge_t* depth = problem->result.depth;
+      
+      auto kernel = [G, depth] __device__ (vertex_t node, queue_t q) -> void {
+          
+          vertex_t d = ((volatile vertex_t * )depth)[node];
+          
+          const vertex_t start  = G.get_starting_edge(node);
+          const vertex_t degree = G.get_number_of_neighbors(node);
+          
+          for(int idx = 0; idx < degree; idx++) {
+              vertex_t neib  = G.get_destination_vertex(start + idx);
+              vertex_t old_d = atomicMin(depth + neib, d + 1);
+              if(old_d > d + 1) {
+                  q.push(neib);
+              }
+          }
+      };
+      // </user-defined>
+      
+      // <boiler-plate>
+      q.launch_thread(numBlock, numThread, kernel);
+      q.sync();
```

Comment:
What I mean by the above comment is this part could be enactor's job.
```cpp
void enact() {
q.launch_thread(numBlock, numThread, loop);
q.sync()
}
```

and the user gets to define:
```cpp
__device__ void loop(vertex_t node, queue_t q) {
...
```

I don't understand it super well to say that the user may not need to control the `launch_thread` and `sync` of the queue.

Reply:
I can try this, though it doesn't exactly follow the pattern of the standard `essentials` loop, which doesn't take arguments.  I suppose the user would want to control `launch_thread` and `sync` if they have multiple lambas that they want to call, but I agree it could be separated and abstracted better.

Reply:
Yeah, let me think more about it then. We can for sure find a way to hide this. I think at the very least, the higher-level enact should have some control, either the kernel launch or the iterations, etc. Not sure what right now tho.